### PR TITLE
B2: petra engine behind the UpdateStrategy trait — schema v2, bitwise parity green

### PR DESCRIPTION
**Pull request by vsletten**
## Summary
Card B2 delivered by a two-worker Hermes relay + supervision closeout — the full RFC-001 implementation:

- **Engine refactor**: the CTMC loop is now `ExactCtmc` behind the `UpdateStrategy` trait with a public `StepCtx` rng/apply seam; other strategies compile-reject at selection (B3's job)
- **Deck schema v2**: `[structure]/[dynamics]/[execution]/[observables]` parse completely, including every RFC surface — `select_mode` (with the exactly-one draw implemented in ExactCtmc as draw 5 per §4.1), per-site source rules, `rekind` init passes, per-strategy `[execution.*]` blocks (parse+validate)
- **v1 compat**: shipped decks load unchanged via the pure load-time shim, tested to full compiled-table equality (not just compile-success)
- **Determinism**: draw order preserved exactly — **bitwise parity on three gates**: kaolinite 20k, kossel 20k (enlarged dims), and the shipped kossel deck's complete 1,512-event natural lifetime (SHA-256 pinned goldens; regeneration moved to an explicit generator so oracles can't be silently overwritten). splitmix64 replica-seed test vectors green.
- **DEVIATION (dispositioned)**: the shipped 8³ kossel deck exhausts at 1,512 events, so full-lifetime parity replaces a meaningless 20k run on it — the stronger claim anyway; enlarged-dims 20k gate retained alongside.

Notable process artifact: worker 1 ran an adversarial self-review that surfaced 7 real findings (public-seam gaps, config drift, missing schema surfaces, oracle-overwrite risk, weak round-trip tests); all were supervision-dispositioned and fixed by worker 2. Card Progress log has the full relay.

## Test plan
- `cargo test --workspace` green (core, deck, parity, schema-v2 suites)
- `cargo clippy --workspace --all-targets -- -D warnings` clean; build clean
- Three bitwise parity gates green; `--paranoid` clean on kaolinite (5k) and kossel (20k, runs to complete dissolution)

## Breaking changes
None — v1 decks load unchanged (tested); CLI behavior preserved.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Refactor Petra’s CTMC engine behind UpdateStrategy, complete schema v2 with v1 compatibility, and lock in deterministic behavior through comprehensive parity gates.

New Features:
- Add the public UpdateStrategy engine seam with ExactCtmc as the initial executable strategy.
- Support complete schema v2 deck parsing and validation, including source rules, selection modes, initialization rekindling, execution strategy blocks, and observables.
- Provide deterministic replica seed policies with pinned splitmix64 behavior.

Bug Fixes:
- Preserve legacy v1 deck behavior through load-time normalization and compiled-table equivalence.
- Prevent parity oracle regeneration during normal tests by requiring an explicit maintenance command.
- Correct source-event handling, per-site runtime kinds, and transition application behavior.

Enhancements:
- Preserve the pre-refactor CTMC draw and update ordering behind the strategy abstraction.
- Fail closed for unsupported grid structures and non-CTMC strategies while allowing their schema surfaces to parse.
- Document and disposition RFC decisions for source semantics, Metropolis timing, observables, disorder representation, and deterministic draws.

Documentation:
- Record completion of B2 and document the resolved schema and determinism decisions in the program status, card, and RFC.

Tests:
- Add migration-grade bitwise parity gates for kaolinite, enlarged Kossel, and the shipped Kossel full lifetime.
- Add schema v2, v1 compatibility, public strategy seam, initialization, source-rule, selection-mode, and seed-vector coverage.

### CODE REVIEW

File: docs/program/cards/B2-engine-refactor.md [hunk 1]
```diff
@@ -20,4 +20,62 @@ strategies in this card (that's B3) — this card is pure refactor.
...
 ## Progress
+
+- 2026-08-20 — hermes worker 1 (sol) — claimed; goldens captured BEFORE
+  refactor; ExactCtmc behind the UpdateStrategy trait; v2 shim;
+  splitmix64 vectors; 20k byte-identical parity; then adversarial
+  self-review surfaced 7 real findings; ceiling hit with precise
+  handoff (incremental-bookkeeping amendment held).
+- 2026-08-20 — hermes worker 2 (sol) — continuation from dirty diff with
+  supervision-dispositioned findings: all 7 addressed. DEVIATION
+  (dispositioned): shipped kossel 8^3 deck exhausts at 1,512 events, so
```
[the remark this answers is not in the document]
Comment (vsletten) [reply docs/program/cards/B2-engine-refactor.md]: ✅ Addressed in 2dabfaa57d6adc8cfbc01a113a083e3414cec2ca: capitalized "kossel" → "Kossel" on line 31 for naming consistency.

**Comment by vsletten** [comment]
👋 PR watcher active. I'll address review comments and CI failures automatically. I submit a formal GitHub review after the automated pass. Labels human-merge and hold block auto-merge; needs-human means I gave up.

**Comment by vsletten** [comment]
🤖 PR watcher review: **APPROVE** (no blocking issues).

Verified against the actual code, not just the PR body:

**Build & tests (real execution, rustc 1.97.1):**
- `cargo test --workspace` → **63 passed, 0 failed** — including all 3 bitwise parity golden gates (kaolinite 20k, kossel 20³ 20k, shipped-kossel full-lifetime 1,512 events) and the 18 schema-v2 tests.
- `cargo build --workspace` → clean.

**Refactor quality — sound:**
- `ExactCtmc` behind `UpdateStrategy` with a public `StepCtx`/`ApplyHandle` seam; the core commits writes and refreshes dirty neighborhoods, so strategies can't bypass the determinism contract. The pre-refactor draw order (site → within-site event → Poisson wait → branch) is preserved exactly — confirmed by the byte-identical goldens.
- Schema v2 deserializer is a clean one-pass shim: schema-less v1 → v2/ctmc, explicit v2 → same dense tables, unknown/mixed schemas fail closed. `select_mode = "one"`, per-site `source` rules, `rekind` init, and per-strategy `[execution.*]` blocks all parse + validate.
- Oracle-overwrite risk correctly mitigated: goldens are read-only in tests; regeneration lives behind an explicit `--regenerate-pinned-oracles` maintenance binary.
- splitmix64 replica-seed vectors match the RFC exactly.

**One non-blocking observation (pre-existing, not introduced here):**
- `cargo clippy --workspace --all-targets -- -D warnings` reports 3 `unnecessary_unwrap` errors on `sel.label` in `petra-core/src/reaction.rs`. I confirmed the identical `sel.label.unwrap()` calls exist on `main` at the same three sites — this is newer-clippy lint drift, not a regression from this PR. The PR body's "clippy clean" claim doesn't hold under rustc 1.97.1, but it's unrelated to the refactor and doesn't block merge.

**Approval note:** PR author is `vsletten` (repo owner), so GitHub forbids self-approval; this summary stands in for the formal review under the owner/collaborator waiver.

**Comment by vsletten** [comment]
🆘 needs-human: unresolvable review — merge blocked by required conversation resolution.

Tried:
- Full automated review → APPROVE, no blocking issues (see 🤖 review comment)
- Verified locally: 63 workspace tests pass, 3 bitwise parity goldens green, build clean
- Confirmed all CI checks pass (Analyze c-cpp / javascript-typescript / python / rust, CodeQL, Sourcery review)
- Attempted merge → rejected by base branch policy (`required_conversation_resolution`)

Blocker: one unresolved Sourcery thread proposes a non-trivial `StepOutcome` API redesign (drop the placeholder `Fired { step: 0, time: 0.0 }` pattern and have the engine construct `Fired`). That's a design change touching the trait, all strategies, call sites, and tests — not a cheap nit, so per policy I'm leaving it for human disposition rather than thrashing the refactor.

Next human step: apply the Sourcery suggestion, resolve the thread as won't-fix, or merge with admin. I will not use `--admin` to bypass the gate.

(Unrelated note: `cargo clippy -D warnings` reports 3 pre-existing `unnecessary_unwrap` errors on `sel.label` in reaction.rs — present on main, newer-clippy lint drift, not introduced by this PR.)

File: petra/crates/petra-core/src/engine.rs [hunk 2]
```diff
@@ -242,160 +440,100 @@ impl Engine {
...
+    pub fn step_with(&mut self, strategy: &mut impl UpdateStrategy) -> Result<StepOutcome, Stop> {
+        let mut pending = Vec::new();
+        let result = {
+            let mut ctx = StepCtx {
+                lattice: &self.lattice,
+                rules: &self.reactions,
+                rng: &mut self.rng,
+                apply: ApplyHandle {
+                    lattice: &self.lattice,
+                    rules: &self.reactions,
```
[the remark this answers is not in the document]
Comment (vsletten) [reply petra/crates/petra-core/src/engine.rs:461]: ℹ️ PR watcher: leaving this one for human. The `step_with` → `StepOutcome` decoupling is a real API change touching `UpdateStrategy`, all strategy impls, and their tests — not a cheap/low-risk mechanical fix, so I'm not applying it automatically. The two high-level notes (quadratic `contains` and boundary-error conflation) are addressed in 909aef2.

**Comment by vsletten** [comment]
🔧 CI fix pushed: addressed Sourcery high-level notes in 909aef2.

- `Engine::commit_transitions`: replaced `Vec<SiteId>` + repeated `contains` with a marking array for O(1) dedup, preserving deterministic insertion order (a `HashSet` would randomize refresh order and break same-seed reproducibility — verified by the kossel determinism test).
- schema v2 `validate_structure`: split boundary-length vs boundary-value validation into distinct, precise error messages.

Left the `step_with`/`StepOutcome` API decoupling for human — it's a non-trivial refactor across `UpdateStrategy` and its impls, not a mechanical fix.

**Comment by vsletten** [comment]
sourcery api redesign suggestion resolved as won't fix. not worth the lift and risk right now.

**Comment by vsletten** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by vsletten** [comment]
🎉 Merged as da3188b7a1932cab5159be5e63920fc28b0136c2. Branch deleted.
